doc/how_to/authentication/guest_users.md

```diff
@@ -0,0 +1,43 @@
+# Allowing Guest Users
+
+Sometimes you will want to allow guest users to have access to certain endpoints of your application or grant them access to all applications without having to log in. For these cases you can make the authentication flow optional and guide them towards the `login_endpoint` only when they are accessing functionality that requires authentication.
+
+## Optional Authentication
+
+The simplest way to make authentication optional is to set the `--oauth-optional` on the commandline or when using `pn.serve` passing the `oauth_optional` keyword argument:
+
+```bash
+panel serve app.py ... --oauth-optional
+```
+
+Panel will now let the user access all endpoints without being redirected to the authentication provider or local login page. If specific functionality in your application requires authentication you may then redirect the user to the [login endpoint](./configuration#Endpoints), e.g. by default you would redirect them to `/login`. As an example let's take this app:
+
+```python
+import panel as pn
+
+pn.extension(template='material')
+
+pn.state.template.title = 'Optional Auth'
+
+if pn.state.user == 'guest':
+    button = pn.widgets.Button(name='Login').servable(target='header')
+    button.js_on_click(code='window.location.href="/login"')
```

Comment:
I am not sure I understand how that works. What if I want to hide everything, shall I just put the `template.servable()` in the `else` (my attempt did not work. Also during my tries I got this error: `2023-10-27 15:24:49,470 DocumentLifeCycleHandler on_session_destroyed callback <bound method OAuthProvider._remove_user of <panel.auth.OAuthProvider object at 0x168aaf220>> failed with following error: 'NoneType' object has no attribute 'decode'
`)? Or do I need to check the user on every pane?

Reply:
That should work fine and does when I try:

```python
import panel as pn

pn.extension(template='material')

pn.state.template.title = 'Optional Auth'


pn.Column(f'# Hello {pn.state.user}!', pn.state.user_info).servable()

if pn.state.user == 'guest':
    button = pn.widgets.Button(name='Login').servable(target='header')
    button.js_on_click(code='window.location.href="/login"')
else:
    pn.Column('My super secret content that I can only see if I am logged in.').servable()
```

<img width="382" alt="Screenshot 2023-10-30 at 12 05 17" src="https://github.com/holoviz/panel/assets/1550771/0dbcfb56-117b-4b42-b012-cba01686c47f">

Reply:
I might be doing something wrong on my side, but I still cannot make it work.

```python
print(pn.state.user)
if pn.state.user == 'guest':
    prompt = pn.pane.Markdown("Login to access App")

    login_button = pn.widgets.Button(name="Login", width=100)
    login_button.js_on_click(code="""window.location.href = './login'""")

    template.main[:, :] = pn.Column(prompt, login_button)

    template.servable()
else:
    template.header.append(...)
    template.main[:, :] = ...
    template.servable()
    ...
```

Here `print(pn.state.user)` starts with None, then goes to "guest" and after I click the button and authenticate with Google, I am back on the same page and it again prints "guest". Did I miss something on the OAuth setup part?

Also when a connection is closed, I get the following:

`2023-10-31 21:41:12,873 DocumentLifeCycleHandler on_session_destroyed callback <bound method OAuthProvider._remove_user of <panel.auth.OAuthProvider object at 0x17cdebd10>> failed with following error: 'NoneType' object has no attribute 'decode'
`

Reply:
Is there any chance you still have an older development version sticking around?

Reply:
I thought that and re-created an env from scratch but got the same results. My env and PATH looks normal too.

Reply:
Thanks, will look into it.

Reply:
Do you have a more complete example? I just tried to reproduce again and couldn't.

Reply:
I will try to make a more minimal example tomorrow.

My app is here otherwise: https://github.com/Simulation-Decomposition/simdec-python/blob/main/panel/app.py

There is a `make serve-oauth` that I modified like so:

```
	PANEL_OAUTH_SCOPE=email panel serve panel/app.py \
		--show \
		--cookie-secret panel_cookie_secret_oauth \
		--oauth-optional \
		--basic-login-template panel/login.html \
		--logout-template panel/logout.html \
		--oauth-provider google \
		--static-dirs _static=docs/_static \
		--reuse-sessions --warm
```

And on the app itself I was trying to add the if/else as above.

Reply:
I confirm that even your example does not work on my side. I am on a fresh mamba env with 1.3.1. The user remains guest.

Reply:
Sorry I haven't checked back in here, could you file an issue so this doesn't fall through the cracks?

Reply:
Sure no problem, done 😃